### ISSUE TITLE
min_data_in_leaf not exactly honoured

### ISSUE THREAD

**Issue by memeplex** [microsoft/LightGBM#3634]
## How you are using LightGBM?

LightGBM component:  Python package

## Environment info

Operating System: MacOS 11.0.1

CPU/GPU model:2,3 GHz Dual-Core Intel Core i5 / Intel Iris Plus Graphics 640 1536 MB

Python version: 3.8.6

---

When I train a model with `min_data_in_leaf = <n>` and then check the output with `trees_to_dataframe()` filtering only leaf nodes (i.e. `df.split_feature.isna()`), many times the `count` column is < `<n>`. Changing `<n>` shows positive correlation with the reported count, but it's not behaving as a hard limit. I'm pretty sure this is caused by [this code](https://github.com/microsoft/LightGBM/blob/0655d67cc1ee28d536a773a6c5d978d38a0e56c1/src/treelearner/serial_tree_learner.cpp#L297):

```
  // no enough data to continue
  if (num_data_in_right_child < static_cast<data_size_t>(config_->min_data_in_leaf * 2)
      && num_data_in_left_child < static_cast<data_size_t>(config_->min_data_in_leaf * 2)) {
```

So unbalanced splits in which a child has count less than `min_data_in_leaf`  still may happen. This also explains the correlation.

If this is expected to work so, then I think the documentation should be made more precise. If not, then it seems a bug to me.

If what I explained above is not enough for you to reproduce or recognize the issue, I may provide a reproducible example latter, but I think the referenced code is quite clear.

**Comment by memeplex** [comment]
EDIT: changed `>= <n>` for `< n` (obviously)

**Comment by jameslamb** [comment]
Thanks @memeplex !

I can see what you mean about how that code you linked could still allow leaf nodes with `num_data < min_data_in_leaf` for very imbalanced splits.

@guolinke could comment on whether this is a bug or we need to update the docs on that parameter to clarify that it just guarantees that *at least one* of the leaf nodes produced by a split will have `num_data >= min_data_in_leaf`.

**Comment by guolinke** [comment]
We have a change since 3.0.0. For better efficiency, we remove the `count` histogram and use `hessian` to approximate `count`. 
Therefore, the data count per leaf is based on the estimation

**Comment by memeplex** [comment]
Thanks for the explanation. The idea is that a larger hessian means more variation in gradients so a bound may prevent overfitting to highly homogeneous subsets of train? I don't think that the count issue is that important, after all it's yet another regularization heuristic, but maybe improving the wording in the documentation might avoid some confusion.

**Comment by memeplex** [comment]
Oh, it's the sum of hessians, not average hessian. In which case it will be correlated to both size and variation. Great.

**Comment by StrikerRUS** [comment]
@jameslamb Would still you mind update the param description?

**Comment by jameslamb** [comment]
sure!

**Comment by jameslamb** [comment]
I've opened #3690 to document this, thanks again for bringing it up @memeplex

**Comment by memeplex** [comment]
Hi, sorry that I wasn't able to timely comment on this. I might be wrong but I find the  note

> this is an approximation based on the hessian, so occasionally you may observe splits which produce leaf nodes that have less than this many observations

misleading. According to the code:

```
  if (num_data_in_right_child < static_cast<data_size_t>(config_->min_data_in_leaf * 2)
      && num_data_in_left_child < static_cast<data_size_t>(config_->min_data_in_leaf * 2)) {
```

the approximation is based on the pre-split num data in a node, not the hessian.

Then there is the suggestion to use `min_sum_hessian_in_leaf` instead, since it can be seen as a lower bound to the product `num_data x avg_hessian` which controls the number of items in the node and the average curvature at the same time. So maybe something like:

> this is an approximation based on the pre-split number of observations, so occasionally you may observe splits which produce leaf nodes that have less than this many observations. You can use min_sum_hessian_in_leaf instead

may be more accurate. I'm not sure, but the suggestion to use min_sum_hessian_in_leaf may even be stronger, almost deprecating the usage of min_data_in_leaf since it's sort of a misnomer at this point.

**Comment by jameslamb** [comment]
@guolinke could you comment on this? It's a little outside of my depth, I think.

**Comment by guolinke** [comment]
@memeplex The code you referred is to pre-filter the nodes that cannot split.
The count approximation is done inside best split finding, you can refer to this line:
https://github.com/microsoft/LightGBM/blob/84c4b7504a49eed63fa819a92db7d0b18292f3f4/src/treelearner/feature_histogram.hpp#L898-L899

**Comment by StrikerRUS** [comment]
Can we close this?

**Comment by memeplex** [comment]
Returning to this. So if I understood well the approximation depends on the assumption that the hessians are more or less uniformly distributed across splits so that `cnt_factor = num_data / sum_hessian` can be used in order to estimate counts as in `hess * cnt_factor`.

The (IMO rather serious) problem I see with the approach is that there is at least a typical case in which it fails in the worst possible way: stratified sampling on the response with inverse probability weighting (as in the Horvitz-Thompson estimator). If hessian = weight = 1 / p, the purest response split will be also the purest weight split, so the hessians will be distributed everything but uniformly across nodes.
